psl-core/src/main/java/org/linqs/psl/reasoner/sgd/term/SGDObjectiveTerm.java

```diff
@@ -87,23 +85,73 @@ public float evaluate(float[] variableValues) {
+        float partial = 0.0f;
+        float dot = dot(variableValues);
 
-        for (int i = 0 ; i < size; i++) {
-            float dot = dot(variableValues);
-            float gradient = computeGradient(i, dot);
-            float gradientStep = gradient * (learningRate / iteration);
+        for (int i = 0; i < size; i++) {
+            partial = computePartial(i, dot);
+            variableStep = computeVariableStep(variableIndexes[i], iteration, learningRate, partial,
+                    accumulatedGradientSquares, accumulatedGradientMean, accumulatedGradientVariance,
+                    adaGrad, adam);
 
-            float newValue = Math.max(0.0f, Math.min(1.0f, variableValues[variableIndexes[i]] - gradientStep));
+            newValue = Math.max(0.0f, Math.min(1.0f, variableValues[variableIndexes[i]] - variableStep));
             movement += Math.abs(newValue - variableValues[variableIndexes[i]]);
             variableValues[variableIndexes[i]] = newValue;
+
+            if (coordinateStep) {
+                dot = dot(variableValues);
+            }
         }
 
         return movement;
     }
 
-    private float computeGradient(int varId, float dot) {
+    private float computeVariableStep(int variableIndex, int iteration, float learningRate, float partial,
+                                      Map<Integer, Float> accumulatedGradientSquares,
+                                      Map<Integer, Float> accumulatedGradientMean,
+                                      Map<Integer, Float> accumulatedGradientVariance,
+                                      boolean adaGrad, boolean adam) {
+        float beta1 = 0.9f * (float)Math.pow(1.0f - 1.0e-8f, iteration - 1);
+        float beta2 = 0.999f;
+        float mean_hat = 0.0f;
+        float variance_hat = 0.0f;
+        float step = 0.0f;
+        float adaptedLearningRate = 0.0f;
+
+        if (adaGrad) {
+            accumulatedGradientSquares.put(variableIndex, accumulatedGradientSquares.getOrDefault(variableIndex, 0.0f)
+                    + (float)Math.pow(partial, 2.0f));
+            adaptedLearningRate = learningRate / (float)Math.sqrt(accumulatedGradientSquares.get(variableIndex) + 1e-8f);
+
+            step = partial * adaptedLearningRate;
+        } else if (adam) {
+            accumulatedGradientMean.put(variableIndex, beta1 * accumulatedGradientMean.getOrDefault(variableIndex, 0.0f)
+                    + (1 - beta1) * partial);
+
+            accumulatedGradientVariance.put(variableIndex, beta2 * accumulatedGradientVariance.getOrDefault(variableIndex, 0.0f) +
+                    (1 - beta2) * (float)Math.pow(partial, 2.0f));
+
+            mean_hat = accumulatedGradientMean.get(variableIndex) / (1 - beta1);
+            variance_hat = accumulatedGradientVariance.get(variableIndex) / (1 - beta2);
+
+            adaptedLearningRate = learningRate / ((float)Math.sqrt(variance_hat) + 1e-8f);
+            step = mean_hat * adaptedLearningRate;
```

Comment:
I need to double check this is correct, can you provide a link to the equations?

Reply:
A good reference for both: https://ruder.io/optimizing-gradient-descent/index.html